packages/keymaps/src/browser/keymaps-service.ts

```diff
@@ -15,85 +15,53 @@
 import { UserStorageUri } from '@theia/userstorage/lib/browser/';
 import parseKeybindings from '../common/keymaps-validation';
 import URI from '@theia/core/lib/common/uri';
 
-export const keymapsUri: URI = new URI('keymaps.json').withScheme(UserStorageUri.SCHEME);
-
 @injectable()
-export class KeymapsService implements Disposable, FrontendApplicationContribution {
-
-    protected readonly toDispose = new DisposableCollection();
-    protected keymapsResource: Resource;
-    protected ready = false;
+export class KeymapsService {
 
     @inject(ResourceProvider)
     protected readonly resourceProvider: ResourceProvider;
 
-    @inject(CommandRegistry)
-    protected readonly commandRegistry: CommandRegistry;
-
     @inject(KeybindingRegistry)
     protected readonly keyBindingRegistry: KeybindingRegistry;
 
     @inject(MessageService)
     protected readonly messageService: MessageService;
 
-    @inject(ILogger)
-    protected readonly logger: ILogger;
+    @inject(OpenerService)
+    protected readonly opener: OpenerService;
+
+    protected resource: Resource;
 
     @postConstruct()
     protected async init() {
-        // this.ajv.compile(keymapsSchema);
-        const resource = await this.resourceProvider(keymapsUri);
-        this.keymapsResource = resource;
-
-        this.toDispose.push(this.keymapsResource);
-
-        if (this.keymapsResource.onDidChangeContents) {
-            this.keymapsResource.onDidChangeContents(() => {
-                this.onDidChangeKeymap();
-            });
+        this.resource = await this.resourceProvider(new URI('keymaps.json').withScheme(UserStorageUri.SCHEME));
+        this.refresh();
+        if (this.resource.onDidChangeContents) {
+            this.resource.onDidChangeContents(() => this.refresh());
         }
-
-        this.parseAndSetKeybindings(await this.keymapsResource.readContents());
-    }
-
-    dispose(): void {
-        this.toDispose.dispose();
-    }
-
-    onStart(app?: FrontendApplication): void {
-        this.ready = true;
     }
 
-    protected onDidChangeKeymap(): void {
-        this.keymapsResource.readContents().then(content => {
-            try {
-                this.parseAndSetKeybindings(content);
-                this.messageService.info(`[${this.keymapsResource.uri}] updated the user keybindings`);
-            } catch (error) {
-                const errors = error.messages || [error.message];
-                for (const message of errors) {
-                    this.messageService.error(`[${this.keymapsResource.uri}] ${message}`);
-                }
+    protected async refresh(): Promise<void> {
+        const content = await this.resource.readContents();
+        try {
+            const keybindings = parseKeybindings(content);
+            this.keyBindingRegistry.setKeymap(KeybindingScope.USER, keybindings);
+        } catch (error) {
+            const errors = error.messages || [error.message];
```

Comment:
in general `error` might be `undefined`, but I haven't checked if it's possible in that case.

Reply:
I did not change this code, let's wait for real issue with it?